# Module 1

   ### Welcome to the Answer notebook for Module 1 ! 
These notebooks have been provided to code and solve all the queries asked in the modules.

This environment has all the necessary libraries pre-installed, and all the Stock, Commodities and Index data files available at the following location - 



#### The problem statements and their corresponding answers are expected to be in the following format - 



#--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------#

### Query 1.1 
Import the csv file of the stock of your choosing using 'pd.read_csv()' function into a dataframe.
Shares of a company can be offered in more than one category. The category of a stock is indicated in the ‘Series’ column. If the csv file has data on more than one category, the ‘Date’ column will have repeating values. To avoid repetitions in the date, remove all the rows where 'Series' column is NOT 'EQ'.
Analyze and understand each column properly.
You'd find the head(), tail() and describe() functions to be immensely useful for exploration. You're free to carry out any other exploration of your own.

In [1]:
import numpy as np 
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
dataset = pd.read_csv('AXISBANK.csv')

In [3]:
dataset

Symbol Series         Date  Prev Close  Open Price  High Price  \
0    AXISBANK     EQ  15-May-2017      503.00      505.00      508.50   
1    AXISBANK     EQ  16-May-2017      500.10      502.10      504.50   
2    AXISBANK     EQ  17-May-2017      501.50      501.35      506.75   
3    AXISBANK     EQ  18-May-2017      502.80      498.00      499.10   
4    AXISBANK     EQ  19-May-2017      492.00      495.75      503.85   
5    AXISBANK     EQ  22-May-2017      501.70      504.00      507.45   
6    AXISBANK     EQ  23-May-2017      502.30      502.25      506.10   
7    AXISBANK     EQ  24-May-2017      500.60      503.00      505.00   
8    AXISBANK     EQ  25-May-2017      499.30      502.00      509.90   
9    AXISBANK     EQ  26-May-2017      507.45      508.00      513.00   
10   AXISBANK     EQ  29-May-2017      511.45      511.90      515.00   
11   AXISBANK     EQ  30-May-2017      507.25      505.00      516.95   
12   AXISBANK     EQ  31-May-2017      514.40      514.00      516.00   
13   AXISBANK     EQ  01-Jun-2017      514.05      512.15      515.95   
14   AXISBANK     EQ  02-Jun-2017      509.30      513.50      513.50   
15   AXISBANK     EQ  05-Jun-2017      508.05      507.35      516.75   
16   AXISBANK     EQ  06-Jun-2017      513.65      514.00      514.40   
17   AXISBANK     EQ  07-Jun-2017      510.20      510.00      515.45   
18   AXISBANK     EQ  08-Jun-2017      514.20      514.00      521.80   
19   AXISBANK     EQ  09-Jun-2017      515.75      515.00      516.35   
20   AXISBANK     EQ  12-Jun-2017      513.65      512.45      513.10   
21   AXISBANK     EQ  13-Jun-2017      507.50      509.00      512.30   
22   AXISBANK     EQ  14-Jun-2017      506.45      509.00      510.90   
23   AXISBANK     EQ  15-Jun-2017      509.50      508.85      509.30   
24   AXISBANK     EQ  16-Jun-2017      507.60      507.65      514.30   
25   AXISBANK     EQ  19-Jun-2017      510.45      511.00      525.35   
26   AXISBANK     EQ  20-Jun-2017      520.15      520.20      521.70   
27   AXISBANK     EQ  21-Jun-2017      511.85      511.05      512.50   
28   AXISBANK     EQ  22-Jun-2017      507.75      508.90      514.45   
29   AXISBANK     EQ  23-Jun-2017      508.00      509.60      510.45   
..        ...    ...          ...         ...         ...         ...   
465  AXISBANK     EQ  27-Mar-2019      757.10      761.30      770.00   
466  AXISBANK     EQ  28-Mar-2019      759.30      761.10      787.65   
467  AXISBANK     EQ  29-Mar-2019      780.70      777.00      781.35   
468  AXISBANK     EQ  01-Apr-2019      777.25      780.95      783.00   
469  AXISBANK     EQ  02-Apr-2019      765.60      769.15      774.95   
470  AXISBANK     EQ  03-Apr-2019      767.40      770.00      772.75   
471  AXISBANK     EQ  04-Apr-2019      761.50      763.00      771.90   
472  AXISBANK     EQ  05-Apr-2019      762.15      763.95      768.75   
473  AXISBANK     EQ  08-Apr-2019      762.20      764.30      768.00   
474  AXISBANK     EQ  09-Apr-2019      754.60      755.00      766.00   
475  AXISBANK     EQ  10-Apr-2019      763.40      761.25      767.90   
476  AXISBANK     EQ  11-Apr-2019      760.85      760.50      761.35   
477  AXISBANK     EQ  12-Apr-2019      752.30      753.40      766.75   
478  AXISBANK     EQ  15-Apr-2019      764.35      764.00      770.20   
479  AXISBANK     EQ  16-Apr-2019      762.85      765.10      776.00   
480  AXISBANK     EQ  18-Apr-2019      771.10      778.40      779.00   
481  AXISBANK     EQ  22-Apr-2019      771.20      768.90      768.90   
482  AXISBANK     EQ  23-Apr-2019      755.50      757.45      759.30   
483  AXISBANK     EQ  24-Apr-2019      753.25      754.70      755.70   
484  AXISBANK     EQ  25-Apr-2019      752.85      754.00      758.55   
485  AXISBANK     EQ  26-Apr-2019      740.90      762.00      765.15   
486  AXISBANK     EQ  30-Apr-2019      759.90      759.95      769.70   
487  AXISBANK     EQ  02-May-2019      766.85      765.40      76

In [4]:
dataset = dataset.groupby('Series').get_group('EQ')

In [5]:
dataset = dataset.reset_index(drop = True)

In [6]:
dataset.head()

Symbol Series         Date  Prev Close  Open Price  High Price  \
0  AXISBANK     EQ  15-May-2017       503.0      505.00      508.50   
1  AXISBANK     EQ  16-May-2017       500.1      502.10      504.50   
2  AXISBANK     EQ  17-May-2017       501.5      501.35      506.75   
3  AXISBANK     EQ  18-May-2017       502.8      498.00      499.10   
4  AXISBANK     EQ  19-May-2017       492.0      495.75      503.85   

   Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
0     498.40      500.25        500.1         502.95                8806756   
1     493.50      502.05        501.5         499.85                8610492   
2     498.50      502.00        502.8         502.88                6443135   
3     490.25      492.85        492.0         494.96                9227022   
4     494.20      501.90        501.7         499.10                7611430   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  
0  4.429384e+09         109965          4320438                   49.06  
1  4.303938e+09         101028          4027982                   46.78  
2  3.240154e+09          84189          2876528                   44.64  
3  4.566993e+09          95624          5348691                   57.97  
4  3.798900e+09          72985          4190505                   55.06

In [7]:
dataset.tail()

Symbol Series         Date  Prev Close  Open Price  High Price  \
489  AXISBANK     EQ  07-May-2019      747.45      751.85      755.50   
490  AXISBANK     EQ  08-May-2019      741.85      741.90      744.75   
491  AXISBANK     EQ  09-May-2019      735.50      733.90      740.50   
492  AXISBANK     EQ  10-May-2019      731.00      734.40      738.50   
493  AXISBANK     EQ  13-May-2019      735.05      734.00      742.85   

     Low Price  Last Price  Close Price  Average Price  Total Traded Quantity  \
489     740.00      744.85       741.85         748.48                5071941   
490     733.35      735.10       735.50         737.58                6739894   
491     729.40      732.35       731.00         734.60                4193451   
492     729.50      732.55       735.05         735.71                3913815   
493     730.00      732.00       732.10         736.74                5891071   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  
489  3.796226e+09         141094          2404026                   47.40  
490  4.971211e+09         118950          3373685                   50.06  
491  3.080497e+09          75621          1409191                   33.60  
492  2.879428e+09         114875          1569735                   40.11  
493  4.340174e+09         142075          2615724                   44.40

In [8]:
dataset.describe()

Prev Close  Open Price  High Price   Low Price  Last Price  \
count  494.000000  494.000000  494.000000  494.000000  494.000000   
mean   577.157389  577.892409  585.013866  570.314170  577.619737   
std     78.697709   79.154912   79.728238   78.278768   78.941719   
min    449.950000  451.650000  457.800000  447.500000  449.800000   
25%    516.250000  516.762500  522.500000  509.425000  515.550000   
50%    545.575000  546.500000  552.325000  537.650000  546.125000   
75%    620.025000  623.412500  628.612500  615.162500  621.087500   
max    780.700000  780.950000  787.650000  769.000000  775.800000   

       Close Price  Average Price  Total Traded Quantity      Turnover  \
count   494.000000     494.000000           4.940000e+02  4.940000e+02   
mean    577.621154     577.686862           8.746869e+06  5.114755e+09   
std      78.934500      78.974233           6.230949e+06  3.714674e+09   
min     449.950000     452.070000           1.245716e+06  6.359149e+08   
25%     516.550000     516.517500           4.955047e+06  2.685941e+09   
50%     545.775000     545.825000           7.130792e+06  4.232753e+09   
75%     620.087500     621.287500           1.027036e+07  6.259857e+09   
max     780.700000     775.220000           5.885946e+07  3.249514e+10   

       No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  
count     494.000000     4.940000e+02              494.000000  
mean   109034.336032     4.198359e+06               47.726802  
std     58759.748772     3.289286e+06               11.953359  
min      7548.000000     3.689650e+05               11.200000  
25%     67426.250000     2.285326e+06               39.732500  
50%     96306.500000     3.251752e+06               46.825000  
75%    135852.500000     5.143684e+06               55.637500  
max    471549.000000     2.551516e+07               86.320000

### Query 1.2
Calculate the maximum, minimum and mean price for the last 90 days. (price=Closing Price unless stated otherwise)

In [9]:
price = []
for i in range(0,len(dataset)):
    if(dataset['Date'][i] == '13-Feb-2019'):
        print(i)
        for j in range(i, len(dataset)):
            price.append(dataset['Close Price'][j])

436


In [10]:
print(price)

[697.5, 695.0, 686.6, 693.3, 693.05, 700.35, 697.2, 702.05, 703.7, 710.15, 718.2, 709.55, 702.4, 732.05, 720.6, 733.75, 733.4, 734.5, 742.55, 738.0, 740.6, 735.1, 751.6, 760.9, 755.75, 757.05, 745.7, 757.1, 759.3, 780.7, 777.25, 765.6, 767.4, 761.5, 762.15, 762.2, 754.6, 763.4, 760.85, 752.3, 764.35, 762.85, 771.1, 771.2, 755.5, 753.25, 752.85, 740.9, 759.9, 766.85, 752.35, 757.3, 747.45, 741.85, 735.5, 731.0, 735.05, 732.1]


In [11]:
maximum_price = max(price)
print(maximum_price)

780.7


In [12]:
minimum_price = min(price)
print(minimum_price)

686.6


In [13]:
mean_price = sum(price)/len(price)
print(mean_price)

740.453448275862


The maximum price over the last 90 days is 780.7. <br>
The minimum price over the last 90 days is 686.6. <br>
The mean price over the last 90 days is 740.453. <br>

### Query 1.3
Analyse the data types for each column of the dataframe. Pandas knows how to deal with dates in an intelligent manner. But to make use of Pandas functionality for dates, you need to ensure that the column is of type 'datetime64(ns)'. Change the date column from 'object' type to 'datetime64(ns)' for future convenience. See what happens if you subtract the minimum value of the date column from the maximum value.

In [14]:
dataset.dtypes

Symbol                     object
Series                     object
Date                       object
Prev Close                float64
Open Price                float64
High Price                float64
Low Price                 float64
Last Price                float64
Close Price               float64
Average Price             float64
Total Traded Quantity       int64
Turnover                  float64
No. of Trades               int64
Deliverable Qty             int64
% Dly Qt to Traded Qty    float64
dtype: object

In [15]:
from functools import partial
to_datetime_fmt = partial(pd.to_datetime, format='%d-%b-%Y')
dataset['Date'] = dataset['Date'].apply(to_datetime_fmt)

In [16]:
dataset.dtypes

Symbol                            object
Series                            object
Date                      datetime64[ns]
Prev Close                       float64
Open Price                       float64
High Price                       float64
Low Price                        float64
Last Price                       float64
Close Price                      float64
Average Price                    float64
Total Traded Quantity              int64
Turnover                         float64
No. of Trades                      int64
Deliverable Qty                    int64
% Dly Qt to Traded Qty           float64
dtype: object

In [17]:
dataset

Symbol Series       Date  Prev Close  Open Price  High Price  \
0    AXISBANK     EQ 2017-05-15      503.00      505.00      508.50   
1    AXISBANK     EQ 2017-05-16      500.10      502.10      504.50   
2    AXISBANK     EQ 2017-05-17      501.50      501.35      506.75   
3    AXISBANK     EQ 2017-05-18      502.80      498.00      499.10   
4    AXISBANK     EQ 2017-05-19      492.00      495.75      503.85   
5    AXISBANK     EQ 2017-05-22      501.70      504.00      507.45   
6    AXISBANK     EQ 2017-05-23      502.30      502.25      506.10   
7    AXISBANK     EQ 2017-05-24      500.60      503.00      505.00   
8    AXISBANK     EQ 2017-05-25      499.30      502.00      509.90   
9    AXISBANK     EQ 2017-05-26      507.45      508.00      513.00   
10   AXISBANK     EQ 2017-05-29      511.45      511.90      515.00   
11   AXISBANK     EQ 2017-05-30      507.25      505.00      516.95   
12   AXISBANK     EQ 2017-05-31      514.40      514.00      516.00   
13   AXISBANK     EQ 2017-06-01      514.05      512.15      515.95   
14   AXISBANK     EQ 2017-06-02      509.30      513.50      513.50   
15   AXISBANK     EQ 2017-06-05      508.05      507.35      516.75   
16   AXISBANK     EQ 2017-06-06      513.65      514.00      514.40   
17   AXISBANK     EQ 2017-06-07      510.20      510.00      515.45   
18   AXISBANK     EQ 2017-06-08      514.20      514.00      521.80   
19   AXISBANK     EQ 2017-06-09      515.75      515.00      516.35   
20   AXISBANK     EQ 2017-06-12      513.65      512.45      513.10   
21   AXISBANK     EQ 2017-06-13      507.50      509.00      512.30   
22   AXISBANK     EQ 2017-06-14      506.45      509.00      510.90   
23   AXISBANK     EQ 2017-06-15      509.50      508.85      509.30   
24   AXISBANK     EQ 2017-06-16      507.60      507.65      514.30   
25   AXISBANK     EQ 2017-06-19      510.45      511.00      525.35   
26   AXISBANK     EQ 2017-06-20      520.15      520.20      521.70   
27   AXISBANK     EQ 2017-06-21      511.85      511.05      512.50   
28   AXISBANK     EQ 2017-06-22      507.75      508.90      514.45   
29   AXISBANK     EQ 2017-06-23      508.00      509.60      510.45   
..        ...    ...        ...         ...         ...         ...   
464  AXISBANK     EQ 2019-03-27      757.10      761.30      770.00   
465  AXISBANK     EQ 2019-03-28      759.30      761.10      787.65   
466  AXISBANK     EQ 2019-03-29      780.70      777.00      781.35   
467  AXISBANK     EQ 2019-04-01      777.25      780.95      783.00   
468  AXISBANK     EQ 2019-04-02      765.60      769.15      774.95   
469  AXISBANK     EQ 2019-04-03      767.40      770.00      772.75   
470  AXISBANK     EQ 2019-04-04      761.50      763.00      771.90   
471  AXISBANK     EQ 2019-04-05      762.15      763.95      768.75   
472  AXISBANK     EQ 2019-04-08      762.20      764.30      768.00   
473  AXISBANK     EQ 2019-04-09      754.60      755.00      766.00   
474  AXISBANK     EQ 2019-04-10      763.40      761.25      767.90   
475  AXISBANK     EQ 2019-04-11      760.85      760.50      761.35   
476  AXISBANK     EQ 2019-04-12      752.30      753.40      766.75   
477  AXISBANK     EQ 2019-04-15      764.35      764.00      770.20   
478  AXISBANK     EQ 2019-04-16      762.85      765.10      776.00   
479  AXISBANK     EQ 2019-04-18      771.10      778.40      779.00   
480  AXISBANK     EQ 2019-04-22      771.20      768.90      768.90   
481  AXISBANK     EQ 2019-04-23      755.50      757.45      759.30   
482  AXISBANK     EQ 2019-04-24      753.25      754.70      755.70   
483  AXISBANK     EQ 2019-04-25      752.85      754.00      758.55   
484  AXISBANK     EQ 2019-04-26      740.90      762.00      765.15   
485  AXISBANK     EQ 2019-04-30      759.90      759.95      769.70   
486  AXISBANK     EQ 2019-05-02      766.85      765.40      767.80   
487  AXISBANK     EQ 2019-05-03      752.35      754.15      763.80   
488  AXISBANK     EQ 2019-05-06

In [18]:
print(max(dataset['Date']) - min(dataset['Date']))

728 days 00:00:00


### Query 1.4
In a separate array , calculate the monthwise VWAP (Volume Weighted Average Price ) of the stock. 
( VWAP = sum(price*volume)/sum(volume) ) 
To know more about VWAP , visit - VWAP definition 
{Hint : Create a new dataframe column ‘Month’. The values for this column can be derived from the ‘Date” column by using appropriate pandas functions. Similarly, create a column ‘Year’ and initialize it. Then use the 'groupby()' function by month and year. Finally, calculate the vwap value for each month (i.e. for each group created).

In [19]:
temp_dates = []
for i in range(0, len(dataset)):
    temp_dates.append(dataset['Date'][i])

month = []
for i in range(0, len(temp_dates)):
    month.append(temp_dates[i].month)

year = []
for i in range(0, len(temp_dates)):
    year.append(temp_dates[i].year)

In [20]:
dataset['Month'] = pd.DataFrame(month)
dataset['Year'] = pd.DataFrame(year)

In [21]:
dataset.head()

Symbol Series       Date  Prev Close  Open Price  High Price  Low Price  \
0  AXISBANK     EQ 2017-05-15       503.0      505.00      508.50     498.40   
1  AXISBANK     EQ 2017-05-16       500.1      502.10      504.50     493.50   
2  AXISBANK     EQ 2017-05-17       501.5      501.35      506.75     498.50   
3  AXISBANK     EQ 2017-05-18       502.8      498.00      499.10     490.25   
4  AXISBANK     EQ 2017-05-19       492.0      495.75      503.85     494.20   

   Last Price  Close Price  Average Price  Total Traded Quantity  \
0      500.25        500.1         502.95                8806756   
1      502.05        501.5         499.85                8610492   
2      502.00        502.8         502.88                6443135   
3      492.85        492.0         494.96                9227022   
4      501.90        501.7         499.10                7611430   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
0  4.429384e+09         109965          4320438                   49.06   
1  4.303938e+09         101028          4027982                   46.78   
2  3.240154e+09          84189          2876528                   44.64   
3  4.566993e+09          95624          5348691                   57.97   
4  3.798900e+09          72985          4190505                   55.06   

   Month  Year  
0      5  2017  
1      5  2017  
2      5  2017  
3      5  2017  
4      5  2017

In [22]:
vwap = []
for i in range(2017, 2020):
    for j in range(1, 13):
        try:
            data = dataset.groupby('Month').get_group(j).groupby('Year').get_group(i)
            vwap.append(str(j) + '-' + str(i) + ': ' + str(round(sum(data['Close Price'] * data['Total Traded Quantity']) / sum(data['Total Traded Quantity']), 2)))
        except:
            continue

In [23]:
print(vwap)

['5-2017: 503.53', '6-2017: 507.95', '7-2017: 521.84', '8-2017: 498.75', '9-2017: 509.1', '10-2017: 487.49', '11-2017: 541.61', '12-2017: 545.12', '1-2018: 588.33', '2-2018: 552.84', '3-2018: 515.75', '4-2018: 524.74', '5-2018: 538.47', '6-2018: 525.44', '7-2018: 537.3', '8-2018: 619.12', '9-2018: 626.03', '10-2018: 572.64', '11-2018: 618.75', '12-2018: 617.29', '1-2019: 667.07', '2-2019: 708.31', '3-2019: 746.13', '4-2019: 760.21', '5-2019: 742.45']


#### The monthwise VWAP is
['5-2017: 503.53',<br>
 '6-2017: 507.95',<br>
 '7-2017: 521.84',<br>
 '8-2017: 498.75',<br>
 '9-2017: 509.1',<br>
 '10-2017: 487.49',<br>
 '11-2017: 541.61',<br>
 '12-2017: 545.12',<br>
 '1-2018: 588.33',<br>
 '2-2018: 552.84',<br>
 '3-2018: 515.75',<br>
 '4-2018: 524.74',<br>
 '5-2018: 538.47',<br>
 '6-2018: 525.44',<br>
 '7-2018: 537.3',<br>
 '8-2018: 619.12',<br>
 '9-2018: 626.03',<br>
 '10-2018: 572.64',<br>
 '11-2018: 618.75',<br>
 '12-2018: 617.29',<br>
 '1-2019: 667.07',<br>
 '2-2019: 708.31',<br>
 '3-2019: 746.13',<br>
 '4-2019: 760.21',<br>
 '5-2019: 742.45']<br>

### Query 1.5
Write a function to calculate the average price over the last N days of the stock price data where N is a user defined parameter. Write a second function to calculate the profit/loss percentage over the last N days.
Calculate the average price AND the profit/loss percentages over the course of last -
1 week, 2 weeks, 1 month, 3 months, 6 months and 1 year.
{Note : Profit/Loss percentage between N days is the percentage change between the closing prices of the 2 days }

In [24]:
def prices_func(N):
    day = max(dataset['Date']) - pd._libs.tslibs.timedeltas.Timedelta(str(N-1) + 'days' + ' 00:00:00')
    index = True
    for i in range(0, len(dataset)):
        if(day == dataset['Date'][i]):
            index = i
            break
    if(index == True):
        day = max(dataset['Date']) - pd._libs.tslibs.timedeltas.Timedelta(str(N) + 'days' + ' 00:00:00')
        for i in range(0, len(dataset)):
            if(day == dataset['Date'][i]):
                index = i
                break
    if(index == True):
        day = max(dataset['Date']) - pd._libs.tslibs.timedeltas.Timedelta(str(N + 1) + 'days' + ' 00:00:00')
        for i in range(0, len(dataset)):
            if(day == dataset['Date'][i]):
                index = i
                break
    prices = []
    for i in range(index, len(dataset)):
        prices.append(dataset['Close Price'][i])
    return(prices)

In [25]:
def mean_func(prices):
    mean_price_n = sum(prices)/len(prices)
    return(mean_price_n)

In [26]:
def profit_or_loss_func(prices):
    profit_or_loss = ((prices[-1] - prices[0])/prices[0])*100
    return(profit_or_loss)

In [27]:
to_find = ['1 week', '2 weeks', '1 month', '3 months', '6 months', '1 year']

In [28]:
day = []
for i in range(0, len(to_find)):
    if('day' in to_find[i]):
        day.append(int(to_find[i][0] + to_find[i][1]))
    if('week' in to_find[i]):
        day.append(int(to_find[i][0] + to_find[i][1]) * 7)
    if('month' in to_find[i]):
        day.append(int(to_find[i][0] + to_find[i][1]) * 30)
    if('year' in to_find[i]):
        day.append(int(to_find[i][0] + to_find[i][1]) * 365)

In [29]:
print(day)

[7, 14, 30, 90, 180, 365]


In [30]:
mean_prices = []
profit_or_loss_perct = []
for i in range(0, len(day)):
    N = day[i]
    prices = prices_func(N)
    mean_prices.append(to_find[i] + ': ' + str(round(mean_func(prices), 2)))
    profit_or_loss_perct.append(to_find[i] + ': ' + str(round(profit_or_loss_func(prices), 2)))

In [31]:
print(mean_prices)

['1 week: 735.1', '2 weeks: 744.38', '1 month: 751.74', '3 months: 740.45', '6 months: 692.07', '1 year: 630.27']


In [32]:
print(profit_or_loss_perct)

['1 week: -1.31', '2 weeks: -4.53', '1 month: -4.22', '3 months: 4.96', '6 months: 16.93', '1 year: 32.65']


#### Average price over the last 
['1 week: 735.1', <br>
'2 weeks: 744.38', <br>
'1 month: 751.74', <br>
'3 months: 740.45', <br>
'6 months: 692.07', <br>
'1 year: 630.27']<br>

#### Profit/Loss percentages over the last
['1 week: -1.31', <br>
'2 weeks: -4.53', <br>
'1 month: -4.22', <br>
'3 months: 4.96', <br>
'6 months: 16.93',<br>
'1 year: 32.65']<br>

### Query 1.6
Add a column 'Day_Perc_Change' where the values are the daily change in percentages i.e. the percentage change between 2 consecutive day's closing prices. Instead of using the basic mathematical formula for computing the same, use 'pct_change()' function provided by Pandas for dataframes. You will note that the first entry of the column will have a ‘Nan’ value. Why does this happen? Either remove the first row, or set the entry to 0 before proceeding.

In [33]:
day_perc_change = dataset['Close Price'].pct_change()

In [34]:
print(day_perc_change)

0           NaN
1      0.002799
2      0.002592
3     -0.021480
4      0.019715
5      0.001196
6     -0.003384
7     -0.002597
8      0.016323
9      0.007883
10    -0.008212
11     0.014096
12    -0.000680
13    -0.009240
14    -0.002454
15     0.011023
16    -0.006717
17     0.007840
18     0.003014
19    -0.004072
20    -0.011973
21    -0.002069
22     0.006022
23    -0.003729
24     0.005615
25     0.019003
26    -0.015957
27    -0.008010
28     0.000492
29    -0.006594
         ...   
464    0.002906
465    0.028184
466   -0.004419
467   -0.014989
468    0.002351
469   -0.007688
470    0.000854
471    0.000066
472   -0.009971
473    0.011662
474   -0.003340
475   -0.011237
476    0.016018
477   -0.001962
478    0.010815
479    0.000130
480   -0.020358
481   -0.002978
482   -0.000531
483   -0.015873
484    0.025644
485    0.009146
486   -0.018909
487    0.006579
488   -0.013007
489   -0.007492
490   -0.008560
491   -0.006118
492    0.005540
493   -0.004013
Name: Close Price, Lengt

In [35]:
dataset['Day_Perc_Change'] = pd.DataFrame(day_perc_change) * 100

In [36]:
dataset['Day_Perc_Change'][0] = 0

In [37]:
dataset.head()

Symbol Series       Date  Prev Close  Open Price  High Price  Low Price  \
0  AXISBANK     EQ 2017-05-15       503.0      505.00      508.50     498.40   
1  AXISBANK     EQ 2017-05-16       500.1      502.10      504.50     493.50   
2  AXISBANK     EQ 2017-05-17       501.5      501.35      506.75     498.50   
3  AXISBANK     EQ 2017-05-18       502.8      498.00      499.10     490.25   
4  AXISBANK     EQ 2017-05-19       492.0      495.75      503.85     494.20   

   Last Price  Close Price  Average Price  Total Traded Quantity  \
0      500.25        500.1         502.95                8806756   
1      502.05        501.5         499.85                8610492   
2      502.00        502.8         502.88                6443135   
3      492.85        492.0         494.96                9227022   
4      501.90        501.7         499.10                7611430   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
0  4.429384e+09         109965          4320438                   49.06   
1  4.303938e+09         101028          4027982                   46.78   
2  3.240154e+09          84189          2876528                   44.64   
3  4.566993e+09          95624          5348691                   57.97   
4  3.798900e+09          72985          4190505                   55.06   

   Month  Year  Day_Perc_Change  
0      5  2017         0.000000  
1      5  2017         0.279944  
2      5  2017         0.259222  
3      5  2017        -2.147971  
4      5  2017         1.971545

### Query 1.7
Add another column 'Trend' whose values are:
'Slight or No change' for 'Day_Perc_Change' in between -0.5 and 0.5 <br>
'Slight positive' for 'Day_Perc_Change' in between 0.5 and 1 <br>
'Slight negative' for 'Day_Perc_Change' in between -0.5 and -1 <br>
'Positive' for 'Day_Perc_Change' in between 1 and 3 <br>
'Negative' for 'Day_Perc_Change' in between -1 and -3 <br>
'Among top gainers' for 'Day_Perc_Change' in between 3 and 7 <br>
'Among top losers' for 'Day_Perc_Change' in between -3 and -7 <br>
'Bull run' for 'Day_Perc_Change' >7 <br>
'Bear drop' for 'Day_Perc_Change' <-7 <br>

In [38]:
trend = []
for i in range(0, len(dataset['Day_Perc_Change'])):
    if((dataset['Day_Perc_Change'][i] < 0.5) & (dataset['Day_Perc_Change'][i] >= -0.5)):
        trend.append('Slight or No change')
    if((dataset['Day_Perc_Change'][i] < 1) & (dataset['Day_Perc_Change'][i] >= 0.5)):
        trend.append('Slight positive')
    if((dataset['Day_Perc_Change'][i] < -0.5) & (dataset['Day_Perc_Change'][i] >= -1)):
        trend.append('Slight negative')
    if((dataset['Day_Perc_Change'][i] < 3) & (dataset['Day_Perc_Change'][i] >= 1)):
        trend.append('Positive')
    if((dataset['Day_Perc_Change'][i] < -1) & (dataset['Day_Perc_Change'][i] >= -3)):
        trend.append('Negative')
    if((dataset['Day_Perc_Change'][i] < 7) & (dataset['Day_Perc_Change'][i] >= 3)):
        trend.append('Among top gainers')
    if((dataset['Day_Perc_Change'][i] < -3) & (dataset['Day_Perc_Change'][i] >= -7)):
        trend.append('Among top losers')
    if(dataset['Day_Perc_Change'][i] >= 7):
        trend.append('Bull run')
    if(dataset['Day_Perc_Change'][i] < -7):
        trend.append('Bear drop')

In [39]:
dataset['Trend'] = pd.DataFrame(trend)

In [40]:
dataset.head()

Symbol Series       Date  Prev Close  Open Price  High Price  Low Price  \
0  AXISBANK     EQ 2017-05-15       503.0      505.00      508.50     498.40   
1  AXISBANK     EQ 2017-05-16       500.1      502.10      504.50     493.50   
2  AXISBANK     EQ 2017-05-17       501.5      501.35      506.75     498.50   
3  AXISBANK     EQ 2017-05-18       502.8      498.00      499.10     490.25   
4  AXISBANK     EQ 2017-05-19       492.0      495.75      503.85     494.20   

   Last Price  Close Price  Average Price  Total Traded Quantity  \
0      500.25        500.1         502.95                8806756   
1      502.05        501.5         499.85                8610492   
2      502.00        502.8         502.88                6443135   
3      492.85        492.0         494.96                9227022   
4      501.90        501.7         499.10                7611430   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
0  4.429384e+09         109965          4320438                   49.06   
1  4.303938e+09         101028          4027982                   46.78   
2  3.240154e+09          84189          2876528                   44.64   
3  4.566993e+09          95624          5348691                   57.97   
4  3.798900e+09          72985          4190505                   55.06   

   Month  Year  Day_Perc_Change                Trend  
0      5  2017         0.000000  Slight or No change  
1      5  2017         0.279944  Slight or No change  
2      5  2017         0.259222  Slight or No change  
3      5  2017        -2.147971             Negative  
4      5  2017         1.971545             Positive

### Query 1.8
Find the average and median values of the column 'Total Traded Quantity' for each of the types of 'Trend'.
{Hint : use 'groupby()' on the 'Trend' column and then calculate the average and median values of the column 'Total Traded Quantity'}

In [41]:
trends = np.array(trend)
trends = np.unique(trends)

In [42]:
average = []
median = []
for i in trends:
        data = dataset.groupby('Trend').get_group(i)
        average.append(i + ': ' + str(round(sum(data['Total Traded Quantity']) / len(data['Total Traded Quantity']), 2)))

In [43]:
print(average)

['Among top gainers: 17643805.65', 'Among top losers: 15525875.11', 'Bear drop: 49308621.0', 'Bull run: 49257708.5', 'Negative: 8020104.35', 'Positive: 9527154.98', 'Slight negative: 7645868.14', 'Slight or No change: 6861540.2', 'Slight positive: 7570645.07']


#### The average total traded quantities for different trends are
['Among top gainers: 17643805.652173914',<br>
 'Among top losers: 15525875.111111112',<br>
 'Bear drop: 49308621.0',<br>
 'Bull run: 49257708.5',<br>
 'Negative: 8020104.3539823005',<br>
 'Positive: 9527154.979166666',<br>
 'Slight negative: 7645868.140625',<br>
 'Slight or No change: 6861540.1953125',<br>
 'Slight positive: 7570645.068965517']<br>

### Query 1.9
SAVE the dataframe with the additional columns computed as a csv file week2.csv. In Module 2, you are going to get familiar with matplotlib, the python module which is used to visualize data.

In [44]:
export_csv = dataset.to_csv (r'week2.csv', index = None, header=True)